### PR TITLE
Add semver CI check 

### PR DESCRIPTION
**Pull request by Morganamilo**
This checks changed files against API surface changes against what the
prdoc says.

It will error if the detected semver change is greater than the one
listed in the prdoc. It will also error if any crates were touched but
not mentioned in the prdoc.

### CODE REVIEW

**Comment by Morganamilo** [comment]
This is finally ready after a lot of headache. See here https://github.com/paritytech/polkadot-sdk/actions/runs/8835275152/job/24259014297?pr=4279 for the output of the action. [b4c1a79](https://github.com/paritytech/polkadot-sdk/pull/4279/commits/b4c1a798887ad5727ab0b1f2aa8be565710deaa6) is just to get some errors to show off the output. It will be reverted before merge.

**Comment by Morganamilo** [comment]
This is just a day after the 1.11 branch off. So it should be feasible to check over the PRdocs that have gotten in before this merge and go with a prdoc based release on the next release.

**Review by EgorPopelyaev — APPROVED** [review]

File: substrate/primitives/staking/src/lib.rs [hunk 1]
```diff
@@ -90,6 +90,7 @@ pub struct Stake<Balance> {
 	/// The total amount of the stash's balance that will be at stake in any forthcoming
 	/// rounds.
 	pub active: Balance,
+	pub major: bool,
```
Comment (ggwpez) [inline substrate/primitives/staking/src/lib.rs:93]: Can you please revert all unrelated changes?

File: .github/workflows/check-semver.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      - name: Rust Cache
+        uses: Swatinem/rust-cache@23bce251a8cd2ffc3c1075eaa2367cf899916d84 # v2.7.3
+        with:
+          cache-on-failure: true
+
+      - name: Rust compilation prerequisites
+        run: |
+          rustup default nightly-2024-03-01
+          rustup target add wasm32-unknown-unknown --toolchain nightly-2024-03-01
+          rustup component add rust-src --toolchain nightly-2024-03-01
```
Comment (ggwpez) [inline .github/workflows/check-semver.yml:29]: We need nightly for some of the SemVer API check i assume?

File: .github/workflows/check-semver.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+      - name: install parity-publish
+        run: cargo install parity-publish@0.5.1
+
+      - name: check semver
+        run: |
+          git config --global --add safe.directory '*'
+          parity-publish --color always prdoc --since ${HEAD}~${COMMITS} --validate prdoc/pr_$PR.prdoc --toolchain nightly-2024-03-01
```
Comment (ggwpez) [inline .github/workflows/check-semver.yml:37]: Maybe you can do a `parity-publish .. || (echo "Help links: .." && exit 1)` or something, so that we can add some custom support links like the [Zepter check](https://github.com/paritytech/polkadot-sdk/blob/f34d8e3cf033e2a22a41b505c437972a5dc83d78/.config/zepter.yaml#L40) does.  

Maybe to the Forum post, once i posted it.

File: substrate/primitives/staking/src/lib.rs (hunk 1 above)
Comment (ggwpez) [reply substrate/primitives/staking/src/lib.rs:93]: Okay this is a test-failure, nvm 👍

**Review by ggwpez — APPROVED** [review]

**Review by ggwpez — CHANGES_REQUESTED** [review]
Blocking until i posted the PSA, so we dont repeat the situation from last time 😅

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/psa-polkadot-sdk-enabling-semver/7616/1

File: .github/workflows/check-semver.yml (hunk 2 above)
Comment (Morganamilo) [reply .github/workflows/check-semver.yml:29]: This works by compiling the crates to rustdoc json output and comparing them. That's an unstable feature so we have to use nightly. Also the format of this json changes so we need to stick to a specific nightly.

**Review by ggwpez — APPROVED** [review]

**Comment by ggwpez** [comment]
Can you suppress compiler warnings? I think that nightly may have more warnings than stable. If it not just warns, but fails to compile, then we have to fix it though.

**Comment by Morganamilo** [comment]
I only just turned them them all because the build was failing to figure out why it was failing after pulling from master. Right now it's all or nothing cause the library isn't very good about controlling it.

I can probably redirect the output to a buffer then dump the buffer on error but focusing on the last curicial bits first.

**Comment by Morganamilo** [comment]
> Can you suppress compiler warnings? I think that nightly may have more warnings than stable. If it not just warns, but fails to compile, then we have to fix it though.

Fixed this by passing rustflags to silence warnings.

File: .github/workflows/check-semver.yml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+          export CARGO_TARGET_DIR=target
+          export RUSTFLAGS='-A warnings -A missing_docs'
+          if ! parity-publish --color always prdoc --since old --validate prdoc/pr_$PR.prdoc --toolchain nightly-2024-03-01 -v; then
+            echo "Help links: https://forum.polkadot.network/t/psa-polkadot-sdk-to-use-semver"
+            exit 1
+          fi
+        env:
+          PR: ${{ github.event.pull_request.number }}
+          HEAD: ${{ github.event.pull_request.head.sha }}
+          COMMITS: ${{ github.event.pull_request.commits }}taplo
```
Comment (ggwpez) [inline .github/workflows/check-semver.yml:48]:
```suggestion
          COMMITS: ${{ github.event.pull_request.commits }}
```
typo?

File: .github/workflows/check-semver.yml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        run: |
+          git config --global --add safe.directory '*'
+          git fetch --no-tags --no-recurse-submodules --depth=1 origin master
+          git branch old origin/master
+
+      - name: check semver
+        run: |
+          export CARGO_TARGET_DIR=target
+          export RUSTFLAGS='-A warnings -A missing_docs'
+          if ! parity-publish --color always prdoc --since old --validate prdoc/pr_$PR.prdoc --toolchain nightly-2024-03-01 -v; then
```
Comment (ggwpez) [inline .github/workflows/check-semver.yml:41]: DQ but where is `old` coming from?

File: .github/workflows/check-semver.yml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            echo "Help links: https://forum.polkadot.network/t/psa-polkadot-sdk-to-use-semver"
```
Comment (ggwpez) [inline .github/workflows/check-semver.yml:42]:
```suggestion
            echo -e "👋 Hello developer! The SemVer information that you declared in the prdoc file did not match what the CI detected.\n\
\n\
Please check the output above and see the following links for more help:\n \
- https://github.com/paritytech/polkadot-sdk/blob/master/docs/contributor/prdoc.md#record-semver-changes\n\
- https://forum.polkadot.network/t/psa-polkadot-sdk-to-use-semver\n\
\n\
Otherwise feel free to ask in the Merge Request or in Matrix chat."
```

It may seem like overkill, but the more information we front-load into the error message the less we have to explain the error.

File: .github/workflows/check-semver.yml (hunk 5 above)
Comment (Morganamilo) [reply .github/workflows/check-semver.yml:41]: The checkout in the git setup above.

File: .github/workflows/check-semver.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
+name: Check semver
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened, ready_for_review]
+    paths:
+      - prdoc/*.prdoc
+
+jobs:
+  check-publish:
```
Comment (ggwpez) [inline .github/workflows/check-semver.yml:10]:
```suggestion
  check-semver:
```

File: .github/workflows/check-semver.yml (hunk 5 above)
Comment (ggwpez) [reply .github/workflows/check-semver.yml:41]: oh yea, sure 🤦

**Review by liamaharon — APPROVED** [review]
